application/frontend/controllers/MfaController.php

```
@@ -0,0 +1,108 @@
+        ]);
+    }
+
+    public function init()
+    {
+        parent::init();
+        $config = \Yii::$app->params['mfa'];
+        $this->idBrokerClient = new IdBrokerClient(
+            $config['baseUrl'],
+            $config['accessToken'],
+            [
+                IdBrokerClient::TRUSTED_IPS_CONFIG              => $config['validIpRanges']       ?? [],
+                IdBrokerClient::ASSERT_VALID_BROKER_IP_CONFIG   => $config['assertValidBrokerIp']   ?? true,
+            ]
+        );
+    }
+
+    public function actionIndex()
+    {
+        try {
+            return $this->idBrokerClient->mfaList(\Yii::$app->user->identity->employee_id);
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+    public function actionCreate()
+    {
+        $type = \Yii::$app->request->getBodyParam('type');
+        if ($type === null) {
+            throw new BadRequestHttpException('Type is required');
+        }
+
+        $label = \Yii::$app->request->getBodyParam('label');
+
+        try {
+            return $this->idBrokerClient->mfaCreate(\Yii::$app->user->identity->employee_id, $type, $label);
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+    public function actionDelete($mfaId)
+    {
+        try {
+            return $this->idBrokerClient->mfaDelete($mfaId, \Yii::$app->user->identity->employee_id);
+        } catch (\Exception $e) {
+            throw $e;
+        }
+    }
+
+    public function actionVerify($mfaId)
+    {
+        $value = \Yii::$app->request->getBodyParam('value');
+        if ($value === null) {
+            throw new BadRequestHttpException('Value is required');
+        }
+
+        try {
+            if ($this->idBrokerClient->mfaVerify($mfaId, \Yii::$app->user->identity->employee_id, $value)) {
+                \Yii::$app->response->statusCode = 204;
+                return;
+            }
+            throw new BadRequestHttpException('Invalid code provided');
```

I think this may be a minor problem (not blocking, but worth addressing after this release). This `BadRequestHttpException` (aka. `HTTP 400`, I think) is immediately caught and instead a `ServerErrorHttpException` (aka. `HTTP 500` is thrown. We may want to move this `throw new BadRequestHttpException('Invalid code provided');` line down below the `catch` block.

---

What about in the catch block starting with:
```php
if ($e instanceof BadRequestHttpException) {
     throw $e;
}
```

---

then we wont log and wont throw the 500 if it was just an invalid code.